### PR TITLE
Improve instructions for how to use this project

### PR DESCRIPTION
**Pull request by jsturtevant**
Updates to the readme to improve instructions on how to use this project.  Tries to address issue #10 

Added two sections:

- one for building this project (boilerplate)
- example for vendoring this project to create own custom metrics api

Couple notes:

- Would like to add a section for how to run and debug. Got it to run locally and connect via kubeconfig file but not sure if I did it correctly (seems that the `RemoteKubeConfigFile` file might not get passed through to the Authentication and Authorization modules correctly. I am going to look at this a bit more then maybe open an issue)
- The example for vendoring just shows what vendoring is (I am new to Go so didn't know this) and which files someone would need to create to get started creating their own.  Would like to show a more complete example of the minimal server/provider implmentations but haven't figured that out yet.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -20,3 +20,6 @@ gofmt:
 	./hack/gofmt-all.sh
 
 verify: verify-gofmt test
+
+run:
```
Comment (DirectXMan12) [inline Makefile]: call this `run-sample` or something, if we have this.
In general, though, while it's possible to run the adapter outside of a pod, it's not going to run out-of-the-box without some flags.  You're generally better of running it in-cluster.

File: README.md [hunk 2]
```diff
@@ -24,10 +24,65 @@ A sample implementation of this can be found in the file `sample-main.go`
 and `pkg/sample-cmd` directory.  You'll want to have the equivalent files
 in your project.
 
-### A note on Dependencies
+### Example for developing your own Custom Metric Server API
+This is an example of how to vendor this project and setup your own provider.  This example uses [glide](https://github.com/Masterminds/glide) but you can use any other [dependency management tool](https://github.com/golang/go/wiki/PackageManagementTools) of your choice.
 
-You'll need to `glide install` dependencies before you can use this
-project.
+#### Create your new project:
```
Comment (DirectXMan12) [inline README.md]: I like this, but perhaps we should move it to a separate file, like `docs/getting-started.md`, and the link here in a clear position so people see it.

File: README.md [hunk 3]
```diff
@@ -24,10 +24,65 @@ A sample implementation of this can be found in the file `sample-main.go`
...
+- `touch pkg/cmd/server/start.go` 
+
+> See the [sample server](https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/master/pkg/sample-cmd/server/start.go) for complete example
+
+#### Add and implement [the custom metrics interface](https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/d8f23423aa1d0ff2bc9656da863d721725b3c68a/pkg/provider/interfaces.go#L84) for your provider:
+
+- `mkdir -p pkg/cmd/provider`
+- `touch pkg/cmd/server/provider.go` 
+
+> See the [sample provider](https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/master/pkg/sample-cmd/provider/provider.go) for complete example
```
Comment (DirectXMan12) [inline README.md]: adding to the above, if we had a separate file, we could go into a bit more detail about why a provided is necessary, and what the server entrypoint is for.

File: README.md [hunk 4]
```diff
@@ -24,10 +24,65 @@ A sample implementation of this can be found in the file `sample-main.go`
...
+
+#### Build the project
+After adding and implementing the files above build you project and run it.
+
+## Development for boilerplate project
+
+### Pre-reqs
+
+- [glide](https://github.com/Masterminds/glide#install) to install dependencies before you can use this project.
+- [Go](https://golang.org/doc/install) 1.8+
```
Comment (DirectXMan12) [inline README.md]: 1.10+ at this point.  Might just want to say "same version of Go as kubernetes" and provide a link, so that we don't have to change it every time kubernetes does (otherwise, in my experience, it gets out of date).

File: README.md [hunk 5]
```diff
@@ -24,10 +24,65 @@ A sample implementation of this can be found in the file `sample-main.go`
...
+### Pre-reqs
+
+- [glide](https://github.com/Masterminds/glide#install) to install dependencies before you can use this project.
+- [Go](https://golang.org/doc/install) 1.8+ 
+- [Mercurial](https://www.mercurial-scm.org/downloads) - one of dependencies requires hg
+- [git](https://git-scm.com/downloads)
+
+### Clone and Build boilerplate project
+
+Clone this repository:
```
Comment (DirectXMan12) [inline README.md]: In general, I'd reword this section a bit.  I realize that not everybody is use to the GitHub workflow, but I'm not sure explaining that is in scope.  Perhaps something like

> There is a sample adapter in this repository that can be used for testing changes to the repository, and also acts as an example implementations.

> To build it, run `make sample-container`, and then `kubectl create -f sample-adapter-deploy`

(and write the `sample-container` make target and the `sample-adapter-deploy` deployment files, which can be based off the ones in the [Prometheus adapter](https://github.com/DirectXMan12/k8s-prometheus-adapter/tree/master/deploy/manifests))

File: Makefile (hunk 1 above)
Comment (jsturtevant) [reply Makefile]: Could we provide an example of how to run it in both scenarios?  I would think running it outside the cluster would make it easier to debug during the development cycle.

What are the flags to run it out-of-the-box?

File: README.md (hunk 2 above)
Comment (jsturtevant) [reply README.md]: sounds good!

File: README.md (hunk 3 above)
Comment (jsturtevant) [reply README.md]: Agreed.  I don't understand those points in detail yet.  Could you add that detail after I swap it out to separate file?

File: README.md (hunk 5 above)
Comment (jsturtevant) [reply README.md]: Makes sense. I will make those updates to include some sample manifests.

File: Makefile (hunk 1 above)
Comment (DirectXMan12) [reply Makefile]: you'd need to pass in kubeconfigs in the right places (you need the main one, plus additional ones for auth, IIRC).

File: README.md (hunk 3 above)
Comment (DirectXMan12) [reply README.md]: sure

**Comment by jsturtevant** [comment]
I started the updates requested.  Still testing them, ran into an issue with glide on my dev box. Will update once I am able to test further.

File: Makefile (hunk 1 above)
Comment (jsturtevant) [reply Makefile]: I tried to do this but didn't i guess I am still missing the correct configs.  Here is what I passed:

` ./_output/amd64/sample-adapter --secure-port=6443  --authentication-kubeconfig="/home/jstur/.kube/config"  --lister-kubeconfig="/home/jstur/.kube/config"`

What am I missing?

**Comment by DirectXMan12** [comment]
Cool, feel free to ping me on slack if you need help with getting things working again.

File: Makefile [hunk 6]
```diff
@@ -1,7 +1,32 @@
...
+ifeq ($(ARCH),arm64)
+	BASEIMAGE?=aarch64/busybox
+endif
+ifeq ($(ARCH),ppc64le)
+	BASEIMAGE?=ppc64le/busybox
+endif
+ifeq ($(ARCH),s390x)
+	BASEIMAGE?=s390x/busybox
+	GOIMAGE=s390x/golang:1.10
+endif
```
Comment (DirectXMan12) [inline Makefile]: For this section, I'd prefer to build from `SCRATCH` and bypass this entirely.

Comment (DirectXMan12) [reply Makefile]: I need to fix that in the other repos I maintain.  We'll need to also mount in `/tmp` as an emptydir in the deployment as well.

File: docs/getting-started.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+## Create the entry point from the sample:
+
+- `touch main.go` 
+
+> See the [sample entry point](https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/master/sample-main.go) for a complete example
+
+## Add the server:
+
+- `mkdir -p pkg/cmd/server`
+- `touch pkg/cmd/server/start.go`
```
Comment (DirectXMan12) [inline docs/getting-started.md]: I wouldn't put this under `pkg/cmd`.  In fact, I'd put main, and this setup, under `cmd/project-name/...`

File: docs/getting-started.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+- `mkdir -p pkg/cmd/server`
+- `touch pkg/cmd/server/start.go` 
+
+> See the [sample server](https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/master/pkg/sample-cmd/server/start.go) for complete example
+
+## Add and implement [the custom metrics interface](https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/d8f23423aa1d0ff2bc9656da863d721725b3c68a/pkg/provider/interfaces.go#L84) for your provider:
+
+- `mkdir -p pkg/cmd/provider`
+- `touch pkg/cmd/server/provider.go`
```
Comment (DirectXMan12) [inline docs/getting-started.md]: just put this under `pkg/provider`.

In general: 
- command-related stuff goes under `cmd`
- actual functionality goes under `pkg`

File: sample-deploy/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,3 @@
+FROM BASEIMAGE
+COPY adapter /
+ENTRYPOINT ["/adapter"]
```
Comment (DirectXMan12) [inline sample-deploy/Dockerfile]: add `--logtostderr` to the entrypoint.  There's basically no condition where you wouldn't want this

File: sample-deploy/manifests/custom-metrics-cluster-role.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: custom-metrics-server-resources
+rules:
+- apiGroups:
+  - custom.metrics.k8s.io
+  resources: ["*"]
+  verbs: ["*"]
```
Comment (DirectXMan12) [inline sample-deploy/manifests/custom-metrics-cluster-role.yaml]: verbs should be `get` and `list`

File: sample-deploy/manifests/custom-metrics-resource-reader-cluster-role.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: custom-metrics-resource-reader
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - namespaces
```
Comment (DirectXMan12) [inline sample-deploy/manifests/custom-metrics-resource-reader-cluster-role.yaml]: add a note that this is the resources that you'll be able to query metrics for

File: Makefile [hunk 12]
```diff
@@ -20,3 +45,14 @@ gofmt:
...
+
+sample-container: build
+	cp sample-deploy/Dockerfile $(TEMP_DIR)
+	cp $(OUT_DIR)/$(ARCH)/sample-adapter $(TEMP_DIR)/adapter
+	cd $(TEMP_DIR) && sed -i "s|BASEIMAGE|scratch|g" Dockerfile
+	sed -i 's|REGISTRY|'${REGISTRY}'|g' sample-deploy/manifests/custom-metrics-apiserver-deployment.yaml
+	docker build -t $(REGISTRY)/$(IMAGE)-$(ARCH):$(VERSION) $(TEMP_DIR)
+	rm -rf $(TEMP_DIR)
+
+sample-push:
```
Comment (DirectXMan12) [inline Makefile]: don't think this is necessary

File: docs/getting-started.md (hunk 8 above)
Comment (jsturtevant) [reply docs/getting-started.md]: see comment above about structure

**Comment by jsturtevant** [comment]
Sorry for the delay.  If the above file structure looks good then I will make the changes requested and get this closed up :-)

File: docs/getting-started.md (hunk 7 above)
Comment (DirectXMan12) [reply docs/getting-started.md]: more or less, yes (start and main don't necessarily have to be separate files, though).

**Comment by DirectXMan12** [comment]
yeah, the file structure looks fine.  In general, `start.go` and `main.go` don't necessarily need to be separate files, though.

**Comment by jsturtevant** [comment]
ready for review

File: README.md [hunk 13]
```diff
@@ -24,10 +24,39 @@ A sample implementation of this can be found in the file `sample-main.go`
 and `pkg/sample-cmd` directory.  You'll want to have the equivalent files
 in your project.
 
-### A note on Dependencies
+### Building your own Custom Metric Server 
 
-You'll need to `glide install` dependencies before you can use this
-project.
+See [getting-started.md](docs/getting-started.md) for a walk through on creating your own custom metric server api.
```
Comment (DirectXMan12) [inline README.md]: nit on wording: "creating your own implementation of the custom metrics API" or "creating you own custom metrics server".

**Comment by jsturtevant** [comment]
Awesome thanks for your help!
